util/fetch/src/client.rs

```
@@ -14,347 +14,554 @@
-			Error::Fetch(ref err) => write!(fmt, "{}", err),
+			Error::Hyper(ref e) => write!(fmt, "{}", e),
+			Error::Url(ref e) => write!(fmt, "{}", e),
+			Error::Io(ref e) => write!(fmt, "{}", e),
+			Error::BackgroundThreadDead => write!(fmt, "background thread gond"),
+			Error::TooManyRedirects => write!(fmt, "too many redirects"),
+			Error::Timeout => write!(fmt, "request timed out"),
+			Error::SizeLimit => write!(fmt, "size limit reached"),
 		}
 	}
 }
 
-impl From<reqwest::Error> for Error {
-	fn from(error: reqwest::Error) -> Self {
-		Error::Fetch(error)
+impl From<hyper::Error> for Error {
+	fn from(e: hyper::Error) -> Self {
+		Error::Hyper(e)
 	}
 }
 
-enum ResponseInner {
-	Response(reqwest::Response),
-	Reader(Box<io::Read + Send>),
-	NotFound,
-}
-
-impl fmt::Debug for ResponseInner {
-	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-		match *self {
-			ResponseInner::Response(ref response) => response.fmt(f),
-			ResponseInner::NotFound => write!(f, "Not found"),
-			ResponseInner::Reader(_) => write!(f, "io Reader"),
-		}
+impl From<io::Error> for Error {
+	fn from(e: io::Error) -> Self {
+		Error::Io(e)
 	}
 }
 
-/// A fetch response type.
-#[derive(Debug)]
-pub struct Response {
-	inner: ResponseInner,
-	abort: Abort,
-	limit: Option<usize>,
-	read: usize,
+impl From<url::ParseError> for Error {
+	fn from(e: url::ParseError) -> Self {
+		Error::Url(e)
+	}
 }
 
-impl Response {
-	/// Creates new successfuly response reading from a file.
-	pub fn from_reader<R: io::Read + Send + 'static>(reader: R) -> Self {
-		Response {
-			inner: ResponseInner::Reader(Box::new(reader)),
-			abort: Abort::default(),
-			limit: None,
-			read: 0,
-		}
+#[cfg(test)]
+mod test {
+	use super::*;
+
+	#[test]
+	fn it_should_fetch() {
+		let client = Client::new().unwrap();
+		let abort = Abort::default().with_max_duration(Duration::from_secs(10));
+		let future = client.fetch("https://httpbin.org/drip?numbytes=3&duration=3&delay=1&code=200", abort);
```

Is there no library to mock responses like this locally so we don't have to have an internet connection to run tests? (And potentially fail tests because `httpbin.org` is down or blocking us or something)

---

I know of [mockito](https://docs.rs/mockito/0.9.1/mockito/) which can replace some of these tests. However it requires `RUST_TEST_THREADS=1` and certain test cases like timeouts are curently not supported AFAICT.

---

Perhaps just spawn an http server in tests? Or at least bind a TcpSocket to a random port and respond with some predefined data.

---

I updated the tests to run a local test server instead.